sqlglot/optimizer/scope.py

```diff
@@ -438,7 +438,22 @@ def traverse(self):
         for child_scope in itertools.chain(
             self.cte_scopes, self.union_scopes, self.table_scopes, self.subquery_scopes
         ):
-            yield from child_scope.traverse()
+            if child_scope.is_union:
+                stack = [(False, child_scope)]
+                scopes = []
+                while stack:
+                    is_right_tree, scope = stack.pop()
+                    if isinstance(scope.expression, exp.Union):
+                        left, right = scope.union_scopes
+                        stack.extend([(True, right), (False, left)])
+                        scopes.append(scope)
+                    else:
+                        yield from scope.traverse()
+                        if is_right_tree:
```

Comment:
why do you only do this for the right tree?

Reply:
should you just add scope to the stack? so it goes right, scope, left ?

Reply:
We need to do yield the left scope tree, then the right, and then the scope. Scope is already in the stack, and I can't just yield from it directly because then the order will be messed up. This check ensures that both subtrees have been visited already. What did you have in mind?

Reply:
i was thinking you could do

stack.extend((right, scope, left)) and then just do the right thing accordingly, rather than store is_right_scope

Reply:
If you add the scope to the stack again won't it keep entering that branch where you extend the stack, resulting in an infinite loop? Maybe if we added another "visited" bool in the stack we could avoid re-entering it, but I guess that defeats the purpose of this simplification, right?